cluster/kubernetes/kubernetes.go

```diff
@@ -483,6 +484,120 @@ func (c *Cluster) PublicSSHKey(regenerate bool) (ssh.PublicKey, error) {
+	}
+
+	// Foreach namespace
+	for _, ns := range namespaces {
+		services, err := c.client.Services(ns.Name).List(api.ListOptions{})
+		if err != nil {
+			return serviceControllers, errors.Wrapf(err, "getting services for namespace %s", ns.Name)
+		}
+
+		controllers, err := c.podControllersInNamespace(ns.Name)
+		if err != nil {
+			return serviceControllers, errors.Wrapf(err, "getting controllers for namespace %s", ns.Name)
+		}
+
+		// Foreach service
+		for _, service := range services.Items {
+			if isAddon(&service) {
+				continue
+			}
+
+			// Find controller for service
+			pc, err := matchController(&service, controllers)
+			if err != nil {
+				c.logger.Log(errors.Wrapf(cluster.ErrNoMatching, "matching controllers to service %s/%s", ns.Name, service.Name))
+				continue
+			}
+			serviceControllers = append(serviceControllers, servicePod{s: service, pc: pc})
+		}
+	}
+	return
+}
+
+// ImagesToFetch is a k8s specific method to get a list of images to update along with their credentials
+func (c *Cluster) ImagesToFetch() func() []registry.ImageCreds {
+	return func() (imageCreds []registry.ImageCreds) {
+		serviceControllers, err := c.allServices("")
+		if err != nil {
+			c.logger.Log("err", errors.Wrapf(err, "fetching images"))
+			return
+		}
+
+		// Foreach service-controller combo
+		for _, servicePod := range serviceControllers {
+			service := servicePod.s
+			controller := servicePod.pc
+			var rawSecrets []v1.LocalObjectReference
+			// If deployment doesn't contain any secrets, just return empty
+			if controller.Deployment != nil && controller.Deployment.Spec.Template.Spec.ImagePullSecrets != nil {
+				rawSecrets = controller.Deployment.Spec.Template.Spec.ImagePullSecrets
+			}
+			creds := registry.NoCredentials()
+			// Foreach secret in PodSpec
+			for _, secName := range rawSecrets {
+				// Get secret
+				sec, err := c.client.Secrets(service.Namespace).Get(secName.Name)
+				if err != nil {
+					c.logger.Log("err", errors.Wrapf(err, "getting secret %q from namespace %q", secName.Name, service.Namespace))
+					continue
+				}
+				if sec.Type != v1.SecretTypeDockercfg {
+					continue
+				}
+				decoded, ok := sec.Data[v1.DockerConfigKey]
+				if !ok {
+					c.logger.Log("err", errors.Wrapf(err, "retrieving pod secret %q", secName.Name))
+					continue
+				}
+
+				// Parse secret
+				crd, err := registry.ParseCredentials(decoded)
+				if err != nil {
+					c.logger.Log("err", err.Error())
+					continue
+				}
+
+				// Merge into the credentials for this PodSpec
+				creds.Merge(crd)
+			}
+
+			// Now create the service and attach the credentials
+			svc := c.makeService(service.Namespace, &service, []podController{controller})
+			for _, ctn := range svc.Containers.Containers {
+				r, err := flux.ParseImageID(ctn.Image)
+				if err != nil {
+					c.logger.Log("err", err.Error())
+					continue
+				}
+				imageCreds = append(imageCreds, registry.ImageCreds{ID: r, Creds: creds})
```

Comment:
At the moment I don't think this is a performance issue and I think it simplifies the code.

First, the only performance hit is memory usage. You're duplicating items in the map. This is very minor.

When the creds are used, the warmer gets the Image-Cred pair and uses them to create a new client. If we weren't using an Image-Cred pair, I.e. some Image->mapToCred type structure, it would still have to create a new client.

So there is no performance or client benefit to not-duplicating.

Finally, I think de-duplicating, and the additional code required in the warmer would significantly increase the complexity of what is very simple. Get a list of image-credential pairs.